src/core/common/settings.hpp

```diff
@@ -1111,6 +1120,8 @@ class Settings : public SettingsBase, private NonCopyable
     otError Save(Key aKey, const void *aValue, uint16_t aSize);
     otError Add(Key aKey, const void *aValue, uint16_t aSize);
     otError Delete(Key aKey);
+
+    const uint16_t mCriticalKeys[3] = {kKeyActiveDataset, kKeyPendingDataset, kKeySrpEcdsaKey};
```

Comment:
@jwhui @abtink Do we encourage in-class member initialization or always initialize in the constructor?

Reply:
How about declaring this as static constant variable, i.e. `static const uint16_t kCritialKeys[]` here? and then define the array values in `cpp` file? This way, we know for sure, the array would be in flash and there would be single instance of it.

The `const` member variable model should work fine but (most) toolchains may not optimize it to use flash (i.e. they would use a variable in RAM and then copy the constant array values to it from constructor). Technically a `const` member variable is saved in RAM and can be initialized from constructor and cannot be changed afterwards.